src/main/java/org/micromanager/acqj/internal/Engine.java

```diff
@@ -379,6 +382,19 @@ private void acquireImages(final AcquisitionEvent event,
                    cameraImageCounts.put(cameraDeviceName, (int) event.getSequence().stream().filter(
                             e -> e.getCameraDeviceName() != null && e.getCameraDeviceName().equals(cameraDeviceName)).count());
                    core_.startSequenceAcquisition(cameraDeviceName, cameraImageCounts.get(cameraDeviceName), 0, true);
+                   // TODO: Run after exposure hooks on a separate thread that checks if
+                   // sequence finished.  This function should probably only exit after this
+                   // thread finishes (Future?)
+                   future = new FutureTask<>(() -> {
+                      while (core_.isSequenceRunning(cameraDeviceName)) {
+                         Thread.sleep(50);
+                      }
+                      for (AcquisitionHook h : event.acquisition_.getAfterExposureHooks()) {
+                         h.run(event);
+                      }
+                      return null;
+                   });
+                   acqExecutor_.execute(future);

```

Comment:
Im a bit confused about this. the acqExecutor_ is a single thread, this code is already executing on in. So I wouldn't think you'd need to invoke it again. I do think you're right that this should happen on a different thread though, since otherwise it is interleaved with copying images out of the circ buffer, and these things should be able to occur independently. Maybe the thing to do here is to launch a new thread, or make a dedicated executor for this type of hook